HDDS-4932. [FSO] Provide list subpaths function to perform recursive ACL check during delete and rename op

rakeshadr opened the pull request:
## What changes were proposed in this pull request?

This task is to provide list subpaths interface to perform recursive ACL check during delete and rename operation.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-4932

## How was this patch tested?

Added UT.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1828,6 +1828,57 @@ public boolean checkAcls(ResourceType resType, StoreType storeType,
+        .setVolumeName(vol)
+        .setBucketName(bucket)
+        .setKeyName(key).build();
+    RequestContext context = RequestContext.newBuilder()
```
It seems every time the RequestContext changes, we will have to add a chain of checkAcls change. Can we pass a RequestContext (or its Builder) instead here for checkAcls to avoid future churns?

rakeshadr replied:
> RequestContext changes, we will have to add a chain of checkAcls change

@xiaoyuyao Are you pointing to a case where user performed delete/rename on a parentDir which has millions of subpaths, now for every subpath ozoneManager needs to invoke checkAcls by creating a `RequestContext context = RequestContext.newBuilder()` ?

Please correct me if I missed something. Thanks!

xiaoyuyao replied:
No I mean we have many checkAcls() with lots of parameters and each time, we add one more, the chain will be extended longer. Instead of doing that, we can pass the wrapper RequestContext or its Builder for checkAcls parameter.

rakeshadr replied:
Thanks @xiaoyuyao for the clear explanation. I will refactor it.

rakeshadr replied:
Done!

xiaoyuyao commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/acl/OzonePrefixPath.java:
```diff
@@ -0,0 +1,59 @@
+   * @return list of immediate files or directories under the given keyPrefix.
+   * @throws IOException
+   */
+  List<OzoneFileStatus> getChildren(String volumeName,
```
It seems that HDDS-4691 should use this API to do the recursion. Also can we provide a batching parameters so that if the number keys are huge, we won't hit RPC packet size limitation or memory limitation.

rakeshadr replied:
Thanks @xiaoyuyao for the review comments. Very good point.

How about to change the List to Iterator ?
```

interface OzonePrefixPath {

   Iterator<? extends OzoneFileStatus> getChildren(String keyPrefix) throws IOException;

   class PathIterator implements Iterator<OzoneFileStatus> {

      @Override
      public boolean hasNext() {
        // batch wise fetching....constant batchSize=1000
        return false;
      }

      @Override
      public OzoneFileStatus next() {
        return null;
      }
    }

  }
```

xiaoyuyao replied:
That looks good to me.

rakeshadr replied:
Done!

xiaoyuyao reviewed (commented):
Thanks @rakeshadr  for working on this. Patch LGTM overall, a few comments added inline...

rakeshadr commented:
@xiaoyuyao , I have addressed the comments. Could you please review it again, thanks!

xiaoyuyao commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/security/acl/RequestContext.java:
```diff
@@ -43,20 +43,23 @@
       ACLIdentityType aclType, ACLType aclRights,
-      String ownerName, boolean recursiveAccessCheck) {
+      String ownerName, boolean recursiveAccessCheck,
+      OzonePrefixPath ozonePrefixPath) {
```
Should we put the OzonePrefixPath in the OzoneObject?

rakeshadr replied:
Thanks @xiaoyuyao for the comment. Agreed.

I hope you are thinking that, OzoneObject has reference to `volume, bucket and keyname` attributes, so we can also add `ozonePrefixPath` to OzoneObject to make it more sense. Am I correct?

Is that `recursiveAccessCheck` flag in RequestContext work instead of adding a new flag to indicates the check for prefixPath? 

Now, `recursiveAccessCheck` flag will be true for any directory and false otherwise. I'm setting the value in OMKeyRenameRequestV1 and OMKeyDeleteRequestV1.

rakeshadr replied:
@xiaoyuyao , I have updated the patch by addressing your comment. Thanks for the reviews!

rakeshadr commented:
Note: Test failure is urelated to the patch, triggered another build run.

```
Warning:  Tests run: 1, Failures: 0, Errors: 0, Skipped: 1, Time elapsed: 0.002 s - in org.apache.hadoop.ozone.client.rpc.TestContainerStateMachineFailureOnRead
[INFO] Running org.apache.hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis
Error:  Tests run: 80, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 62.65 s <<< FAILURE! - in org.apache.hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis
Error:  testValidateBlockLengthWithCommitKey  Time elapsed: 0.019 s  <<< FAILURE!
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
```

rakeshadr commented:
@xiaoyuyao test case failure is unrelated to the patch. I've raised jira HDDS-5012 to address it.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:
```diff
@@ -186,6 +189,64 @@ public void checkAcls(OzoneManager ozoneManager,
+    boolean isDirectory = pathViewer.getOzoneFileStatus().isDirectory();
+
+    RequestContext.Builder contextBuilder = RequestContext.newBuilder()
+        .setAclRights(IAccessAuthorizer.ACLType.DELETE)
```
why do we hard code ACLType to DELETE here?

rakeshadr replied:
Sure, will pass `ACLType` as an argument to this function and will make it generic.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:
```diff
@@ -607,11 +609,20 @@ public static OzoneFileStatus getOMKeyInfoIfExists(
     String bucketKey = omMetadataMgr.getBucketKey(volumeName, bucketName);
     OmBucketInfo omBucketInfo =
             omMetadataMgr.getBucketTable().get(bucketKey);
+    if (null == omBucketInfo) {
```
Does validateBucket handle the non-exist bucket case?

rakeshadr replied:
OK, will remove this. Its a duplicate check.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyDeleteRequestV1.java:
```diff
@@ -94,16 +91,13 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
-      // check Acl
-      checkKeyAcls(ozoneManager, volumeName, bucketName, keyName,
-          IAccessAuthorizer.ACLType.DELETE, OzoneObj.ResourceType.KEY);
+      checkACLs(ozoneManager, volumeName, bucketName, keyName);
```
If this is intended for delete only, should we rename it to checkACLsForDelete?

rakeshadr replied:
Same as above. I will pass `ACLType` as an argument to this function. Hope this make sense to you.

rakeshadr commented:
Thanks a lot @xiaoyuyao for the useful review comments. Updated PR by addressing the comments.

xiaoyuyao commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzonePrefixPathImpl.java:
```diff
@@ -0,0 +1,153 @@
+          currentIterator =
+              getNextListOfKeys(currentValue.getTrimmedName()).iterator();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
```
Can we avoid throw RuntimeException here?

rakeshadr replied:
OK. Then I will return false if any exception thrown by `getNextListOfKeys(currentValue.getTrimmedName())`  function call.

xiaoyuyao reviewed (commented):
Thanks @rakeshadr  for the update, just have one minor comments inline.

rakeshadr commented:
> Thanks @rakeshadr for the update, just have one minor comments inline.

Thanks a lot @xiaoyuyao  for the review comments. Sure, hope I can proceed to merge the patch once that is handled.

rakeshadr commented:
Updated the patch by addressing the comment. Thank you @xiaoyuyao!

rakeshadr commented:
Thanks again @xiaoyuyao for the great support and the review comments helped a lot:-)
I've addressed those comments and got a clean QA report.

I would like to merge this PR today sometime(IST) and I'm planning to rebase the feature branch to make the branch maintenance easy as it is 27 commits behind master. I hope you agree to this idea.

Please feel free to comments(if any), I'm really happy to work on those and make the changes.

xiaoyuyao commented:
+1. Thanks @rakeshadr  for the update, LGTM.

rakeshadr commented:
> +1. Thanks @rakeshadr for the update, LGTM.

Thanks @xiaoyuyao
